File: scripts/life-agent/assign-wechat-avatars.ts
/**
 * 为人生 Agent 分配不重复的「微信风格」头像（下载到本地 uploads，写入 cover_image_url）。
 *
 * 头像来源（按优先级尝试，直到拿到不重复图片）：
 *   1. v2.xxapi.cn 随机头像 API（国内常见 QQ/微信用户头像风格）
 *   2. DiceBear 卡通头像（seed = agentId，微信里也很常见）
 *   3. randomuser.me 人像（兜底）
 *
 * 用法：
 *   npx tsx scripts/life-agent/assign-wechat-avatars.ts          # 预览
 *   npx tsx scripts/life-agent/assign-wechat-avatars.ts --apply  # 写入 DB + 保存文件
 *   npx tsx scripts/life-agent/assign-wechat-avatars.ts --podcast-only --apply  # 仅四批播客（59 条）
 *   LIMIT=20 npx tsx scripts/life-agent/assign-wechat-avatars.ts --apply
 */
import "dotenv/config";
import crypto from "node:crypto";
import fs from "node:fs/promises";
import path from "node:path";
import { PrismaClient } from "@prisma/client";

const prisma = new PrismaClient();

const APPLY = process.argv.includes("--apply") || process.env.APPLY === "1";
const PODCAST_ONLY =
  process.argv.includes("--podcast-only") || process.env.PODCAST_ONLY === "1";
const LIMIT = Number(process.env.LIMIT || 0);
const PODCAST_SOURCES = [
  "不止大学播客",
  "我下班了播客",
  "校招飞播客",
  "迷你退休播客",
] as const;
const FETCH_TIMEOUT_MS = Number(process.env.AVATAR_FETCH_TIMEOUT_MS || 25000);
const LOCAL_COVER_DIR =
  process.env.LIFE_AGENT_COVER_DIR ||
  path.resolve(process.cwd(), "backend", "uploads", "life-agent-covers");

const DICEBEAR_STYLES = [
  "lorelei",
  "micah",
  "adventurer",
  "avataaars",
  "fun-emoji",
  "notionists",
  "personas",
  "big-smile",
] as const;

type AgentRow = {
  id: string;
  displayName: string;
  coverImageUrl: string | null;
};

function sha256(buf: Buffer): string {
  return crypto.createHash("sha256").update(buf).digest("hex");
}

function hashPick(seed: string, mod: number): number {
  const h = crypto.createHash("sha256").update(seed).digest();
  return Number(h.readBigUInt64BE(0) % BigInt(mod));
}

async function fetchWithTimeout(url: string, init?: RequestInit): Promise<Response> {
  return fetch(url, {
    ...init,
    signal: AbortSignal.timeout(FETCH_TIMEOUT_MS),
  });
}

async function tryXxapiHeadUrl(): Promise<string | null> {
  try {
    const res = await fetchWithTimeout("https://v2.xxapi.cn/api/head?return=json");
    if (!res.ok) return null;
    const json = (await res.json()) as { data?: unknown };
    return typeof json.data === "string" && json.data.startsWith("http") ? json.data : null;
  } catch {
    return null;
  }
}

function dicebearUrl(agentId: string, attempt: number): string {
  const style = DICEBEAR_STYLES[hashPick(`${agentId}:${attempt}`, DICEBEAR_STYLES.length)];
  return `https://api.dicebear.com/9.x/${style}/png?seed=${encodeURIComponent(`wx-${agentId}-${attempt}`)}&size=400`;
}

function randomUserUrl(agentId: string, attempt: number): string {
  const gender = hashPick(`${agentId}:${attempt}:g`, 2) === 0 ? "women" : "men";
  const idx = hashPick(`${agentId}:${attempt}:i`, 100);
  return `https://randomuser.me/api/portraits/${gender}/${idx}.jpg`;
}

function sourceCandidates(agentId: string): string[] {
  const out: string[] = [];
  // xxapi 每次请求 URL 不同，在 downloadUniqueAvatar 里单独拉
  for (let attempt = 0; attempt < 8; attempt += 1) {
    out.push(dicebearUrl(agentId, attempt));
    out.push(randomUserUrl(agentId, attempt));
  }
  return out;
}

async function downloadBytes(url: string): Promise<Buffer | null> {
  try {
    const res = await fetchWithTimeout(url, {
      headers: { Accept: "image/*", "User-Agent": "BrightAgentAvatarBot/1.0" },
    });
    if (!res.ok) return null;
    const ct = res.headers.get("content-type") || "";
    if (!ct.startsWith("image/")) return null;
    const buf = Buffer.from(await res.arrayBuffer());
    if (buf.length < 256 || buf.length > 2 * 1024 * 1024) return null;
    return buf;
  } catch {
    return null;
  }
}

async function downloadUniqueAvatar(
  agentId: string,
  usedHashes: Set<string>,
): Promise<{ bytes: Buffer; ext: string; source: string } | null> {
  // 优先：xxapi 随机微信/QQ 风格头像（最多试 6 次）
  for (let i = 0; i < 6; i += 1) {
    const headUrl = await tryXxapiHeadUrl();
    if (!headUrl) continue;
    const bytes = await downloadBytes(headUrl);
    if (!bytes) continue;
    const digest = sha256(bytes);
    if (usedHashes.has(digest)) continue;
    usedHashes.add(digest);
    return { bytes, ext: ".jpg", source: headUrl };
  }

  for (const url of sourceCandidates(agentId)) {
    const bytes = await downloadBytes(url);
    if (!bytes) continue;
    const digest = sha256(bytes);
    if (usedHashes.has(digest)) continue;
    usedHashes.add(digest);
    const ext = url.includes(".jpg") ? ".jpg" : ".png";
    return { bytes, ext, source: url };
  }

  return null;
}

function coverApiPath(filename: string): string {
  return `/api/upload/life-agent-cover/${filename}`;
}

async function main() {
  await fs.mkdir(LOCAL_COVER_DIR, { recursive: true });

  const rows = await prisma.lifeAgentProfile.findMany({
    select: { id: true, displayName: true, coverImageUrl: true },
    where: PODCAST_ONLY ? { source: { in: [...PODCAST_SOURCES] } } : undefined,
    orderBy: { createdAt: "asc" },
    ...(LIMIT > 0 ? { take: LIMIT } : {}),
  });

  console.log(`Agents to process: ${rows.length}`);
  console.log(`Podcast only: ${PODCAST_ONLY}`);
  console.log(`Mode: ${APPLY ? "APPLY (write files + DB)" : "DRY RUN"}`);
  console.log(`Cover dir: ${LOCAL_COVER_DIR}`);

  const usedHashes = new Set<string>();
  let ok = 0;
  let fail = 0;

  for (const row of rows) {
    const avatar = await downloadUniqueAvatar(row.id, usedHashes);
    if (!avatar) {
      fail += 1;
      console.log(`FAIL ${row.displayName} (${row.id}) — no unique avatar downloaded`);
      continue;
    }

    const filename = `wx-${row.id.replace(/-/g, "").slice(0, 24)}${avatar.ext}`;
    const apiPath = coverApiPath(filename);
    ok += 1;

    console.log(`OK   ${row.displayName} -> ${apiPath} [${avatar.source.slice(0, 72)}…]`);

    if (!APPLY) continue;

    await fs.writeFile(path.join(LOCAL_COVER_DIR, filename), avatar.bytes);
    await prisma.lifeAgentProfile.update({
      where: { id: row.id },
      data: {
        coverImageUrl: apiPath,
        coverPresetKey: null,
      },
    });
  }

  console.log("");
  console.log(`Done. success=${ok}, failed=${fail}, unique_images=${usedHashes.size}`);
  if (!APPLY) {
    console.log("Dry run only. Re-run with: npm run assign:wechat-avatars -- --apply");
  }
}

main()
  .catch((error) => {
    console.error(error);
    process.exitCode = 1;
  })
  .finally(async () => {
    await prisma.$disconnect();
  });

File: backend/scripts/assign_wechat_avatars.go
// 为所有人生 Agent 下载不重复的「微信风格」头像到 uploads，并写入 cover_image_url。
//
// 在 backend 目录执行（需 DATABASE_URL，与 seed 相同）：
//
//	go run ./scripts/assign_wechat_avatars.go
//	go run ./scripts/assign_wechat_avatars.go --apply
//	LIMIT=20 go run ./scripts/assign_wechat_avatars.go --apply
//
// 生产 Docker 宿主机示例（~/regr）：
//
//	cd backend && go run ./scripts/assign_wechat_avatars.go --apply
package main

import (
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"io"
	"log"
	"net/http"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"

	"github.com/agent-marketplace/backend/internal/db"
	"github.com/agent-marketplace/backend/internal/models"
	"github.com/joho/godotenv"
)

const (
	maxAvatarBytes = 2 * 1024 * 1024
	minAvatarBytes = 256
)

var dicebearStyles = []string{
	"lorelei", "micah", "adventurer", "avataaars", "fun-emoji", "notionists", "personas", "big-smile",
}

type xxapiResp struct {
	Data string `json:"data"`
}

func main() {
	_ = godotenv.Load()
	_ = godotenv.Load("../.env")

	apply := false
	for _, arg := range os.Args[1:] {
		if arg == "--apply" {
			apply = true
		}
	}

	limit := 0
	if v := strings.TrimSpace(os.Getenv("LIMIT")); v != "" {
		if n, err := strconv.Atoi(v); err == nil && n > 0 {
			limit = n
		}
	}

	coverDir := strings.TrimSpace(os.Getenv("LIFE_AGENT_COVER_DIR"))
	if coverDir == "" {
		coverDir = filepath.Join(".", "uploads", "life-agent-covers")
	}
	if apply {
		if err := os.MkdirAll(coverDir, 0o755); err != nil {
			log.Fatal("mkdir:", err)
		}
	}

	dsn, err := db.DSNFromEnv()
	if err != nil {
		log.Fatal("dsn:", err)
	}
	if err := db.Init(dsn); err != nil {
		log.Fatal("db init:", err)
	}

	var profiles []models.LifeAgentProfile
	q := db.DB.Order("created_at ASC")
	if limit > 0 {
		q = q.Limit(limit)
	}
	if err := q.Find(&profiles).Error; err != nil {
		log.Fatal("query:", err)
	}

	client := &http.Client{Timeout: 25 * time.Second}
	used := make(map[string]struct{})
	ok, fail := 0, 0

	fmt.Printf("agents=%d mode=%s coverDir=%s\n", len(profiles), modeLabel(apply), coverDir)

	for _, p := range profiles {
		img, source, ext, err := downloadUniqueAvatar(client, p.ID, used)
		if err != nil {
			fail++
			log.Printf("FAIL %s (%s): %v", p.DisplayName, p.ID, err)
			continue
		}
		ok++

		filename := fmt.Sprintf("wx-%s%s", strings.ReplaceAll(p.ID, "-", "")[:24], ext)
		apiPath := "/api/upload/life-agent-cover/" + filename
		fmt.Printf("OK   %s -> %s [%s]\n", p.DisplayName, apiPath, truncate(source, 72))

		if !apply {
			continue
		}

		fullPath := filepath.Join(coverDir, filename)
		if err := os.WriteFile(fullPath, img, 0o644); err != nil {
			log.Printf("write %s: %v", fullPath, err)
			fail++
			ok--
			continue
		}
		res := db.DB.Model(&models.LifeAgentProfile{}).Where("id = ?", p.ID).Updates(map[string]interface{}{
			"cover_image_url":  apiPath,
			"cover_preset_key": nil,
		})
		if res.Error != nil {
			log.Printf("update %s: %v", p.ID, res.Error)
		}
	}

	fmt.Printf("\ndone success=%d failed=%d unique_images=%d\n", ok, fail, len(used))
	if !apply {
		fmt.Println("dry run only. Re-run with: go run ./scripts/assign_wechat_avatars.go --apply")
	}
}

func modeLabel(apply bool) string {
	if apply {
		return "APPLY"
	}
	return "DRY_RUN"
}

func truncate(s string, n int) string {
	if len(s) <= n {
		return s
	}
	return s[:n] + "…"
}

func digest(buf []byte) string {
	sum := sha256.Sum256(buf)
	return hex.EncodeToString(sum[:])
}

func hashPick(seed string, mod int) int {
	sum := sha256.Sum256([]byte(seed))
	n := uint64(sum[0])<<56 | uint64(sum[1])<<48 | uint64(sum[2])<<40 | uint64(sum[3])<<32 |
		uint64(sum[4])<<24 | uint64(sum[5])<<16 | uint64(sum[6])<<8 | uint64(sum[7])
	if mod <= 0 {
		return 0
	}
	return int(n % uint64(mod))
}

func downloadUniqueAvatar(client *http.Client, agentID string, used map[string]struct{}) ([]byte, string, string, error) {
	for i := 0; i < 6; i++ {
		url, err := fetchXxapiHeadURL(client)
		if err != nil {
			continue
		}
		buf, err := downloadImage(client, url)
		if err != nil {
			continue
		}
		d := digest(buf)
		if _, dup := used[d]; dup {
			continue
		}
		used[d] = struct{}{}
		return buf, url, extFromURL(url, ".jpg"), nil
	}

	for attempt := 0; attempt < 8; attempt++ {
		for _, url := range []string{
			dicebearURL(agentID, attempt),
			randomUserURL(agentID, attempt),
		} {
			buf, err := downloadImage(client, url)
			if err != nil {
				continue
			}
			d := digest(buf)
			if _, dup := used[d]; dup {
				continue
			}
			used[d] = struct{}{}
			return buf, url, extFromURL(url, ".png"), nil
		}
	}
	return nil, "", "", fmt.Errorf("no unique avatar downloaded")
}

func fetchXxapiHeadURL(client *http.Client) (string, error) {
	req, err := http.NewRequest(http.MethodGet, "https://v2.xxapi.cn/api/head?return=json", nil)
	if err != nil {
		return "", err
	}
	req.Header.Set("User-Agent", "BrightAgentAvatarBot/1.0")
	resp, err := client.Do(req)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return "", fmt.Errorf("xxapi status %d", resp.StatusCode)
	}
	var out xxapiResp
	if err := json.NewDecoder(resp.Body).Decode(&out); err != nil {
		return "", err
	}
	if !strings.HasPrefix(out.Data, "http") {
		return "", fmt.Errorf("xxapi bad data")
	}
	return out.Data, nil
}

func dicebearURL(agentID string, attempt int) string {
	style := dicebearStyles[hashPick(agentID+":"+strconv.Itoa(attempt), len(dicebearStyles))]
	seed := fmt.Sprintf("wx-%s-%d", agentID, attempt)
	return fmt.Sprintf("https://api.dicebear.com/9.x/%s/png?seed=%s&size=400", style, seed)
}

func randomUserURL(agentID string, attempt int) string {
	gender := "women"
	if hashPick(agentID+":"+strconv.Itoa(attempt)+":g", 2) == 1 {
		gender = "men"
	}
	idx := hashPick(agentID+":"+strconv.Itoa(attempt)+":i", 100)
	return fmt.Sprintf("https://randomuser.me/api/portraits/%s/%d.jpg", gender, idx)
}

func downloadImage(client *http.Client, url string) ([]byte, error) {
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Accept", "image/*")
	req.Header.Set("User-Agent", "BrightAgentAvatarBot/1.0")
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("status %d", resp.StatusCode)
	}
	ct := resp.Header.Get("Content-Type")
	if !strings.HasPrefix(ct, "image/") {
		return nil, fmt.Errorf("not image: %s", ct)
	}
	buf, err := io.ReadAll(io.LimitReader(resp.Body, maxAvatarBytes+1))
	if err != nil {
		return nil, err
	}
	if len(buf) < minAvatarBytes || len(buf) > maxAvatarBytes {
		return nil, fmt.Errorf("bad size %d", len(buf))
	}
	return buf, nil
}

func extFromURL(url, fallback string) string {
	lower := strings.ToLower(url)
	switch {
	case strings.Contains(lower, ".jpg"), strings.Contains(lower, ".jpeg"):
		return ".jpg"
	case strings.Contains(lower, ".webp"):
		return ".webp"
	case strings.Contains(lower, ".png"):
		return ".png"
	default:
		return fallback
	}
}
